### PR TITLE
[MRG+1] Datasets: Downloader for COBRE data

### PR DESCRIPTION
**Pull request by KamalakerDadi**
- Added function ```fetch_cobre_niak```, documentation, tests, reference

This function downloads current version of datasets  from download link as posted in Issue #915 

Still to add: ```whatsnew``` and any comments or improvements.

### CODE REVIEW

File: nilearn/datasets/func.py [hunk 1]
```diff
@@ -1411,3 +1412,98 @@ def fetch_megatrawls_netmats(dimensionality=100, timeseries='eigen_regression',
...
+        if n_subjects < max_subjects:
+            # shuffle datasets to have almost equal balance between sch vs ctrl
+            ids = shuffle(ids, random_state=0, n_samples=n_subjects)
+        elif n_subjects > max_subjects:
+            warnings.warn('Warning: there are only %d subjects' % max_subjects)
+            n_subjects = max_subjects
+    else:
+        n_subjects = max_subjects
+
+    func_filenames = [('fmri_' + i.strip(' "\'') +
```
Comment (aabadie) [inline nilearn/datasets/func.py]: This line and the one above doesn't work on python 3, you should use `decode()` in `i` because it's a byte string :
```python
func_filenames = [('fmri_' + i.decode().strip(' "\'') 
```

Comment (KamalakerDadi) [inline nilearn/datasets/func.py]: Thanks. I thought its ```encode()```. Is it not ?

Comment (KamalakerDadi) [inline nilearn/datasets/func.py]: Could you point me any documentation to find this difference ? If you have any ?

Comment (aabadie) [inline nilearn/datasets/func.py]: I'm not sure it will work work with `encode()`. `np.recfromcsv` returns an array with `numpy._bytes`. Those only have a `decode()` method.
You can see in the code another place with the same trick : https://github.com/nilearn/nilearn/blob/master/nilearn/datasets/func.py#L1155
where `pheno` contains the result of `np.recfromcsv`.

**Comment by GaelVaroquaux** [comment]
Maybe the function should be called "fetch_cobre", and we'll later add an
option to decide between different preprocessing (niak or otherwise).

**Comment by KamalakerDadi** [comment]
Can somebody review this PR ? As we need this for Brain Hack. Thanks

**Comment by aabadie** [comment]
LGTM +1, The failing test on travis is not related to your PR (#986 was entered for this). Once CircleCI is done, we can merge this.

File: nilearn/datasets/description/cobre.rst [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+median volume of the first run as a target. The median volume of one selected
+fMRI run for each subject was coregistered with a T1 individual scan using
+Minctracc (Collins and Evans, 1998), which was itself non-linearly transformed
+to the Montreal Neurological Institute (MNI) template (Fonov et al., 2011)
+using the CIVET pipeline (Ad-Dabbagh et al., 2006). The MNI  symmetric
+template was generated from the ICBM152 sample of 152 young adults, after 40
+iterations of non-linear coregistration. The rigid-body
+transform, fMRI-to-T1 transform and T1-to-stereotaxic transform were all
+combined, and the functional volumes were resampled in the MNI space at a 3 mm
+isotropic resolution. The "scrubbing" method of (Power et al., 2012), was used
```
Comment (AlexandreAbraham) [inline nilearn/datasets/description/cobre.rst]: I thought that we had a version without scrubbing?

File: nilearn/datasets/func.py [hunk 3]
```diff
@@ -1411,3 +1412,98 @@ def fetch_megatrawls_netmats(dimensionality=100, timeseries='eigen_regression',
...
+        - 'func': string list
+            Paths to Nifti images.
+        - 'mat_files': string list
+            Paths to .mat files of each subject.
+        - 'phenotypic': ndarray
+            Contains data of clinical variables, sex, age, FD.
+        - 'description': data description of the release and references.
+    """
+    if url is None:
+        url = "https://ndownloader.figshare.com/articles/1160600/versions/15"
```
Comment (AlexandreAbraham) [inline nilearn/datasets/func.py]: The individual files are available, why do we rely on an aggregation of them?

File: nilearn/datasets/func.py [hunk 4]
```diff
@@ -1411,3 +1412,98 @@ def fetch_megatrawls_netmats(dimensionality=100, timeseries='eigen_regression',
...
+    # Load file in filename to numpy arrays
+    names = ['subject_type', 'sz', 'age', 'sex', 'fd']
+    csv_array = np.recfromcsv(csv_file[0], names=names, skip_header=True)
+    # Get the ids of the datasets
+    ids = csv_array['subject_type']
+    max_subjects = len(ids)
+    if n_subjects is not None:
+        if n_subjects < max_subjects:
+            # shuffle datasets to have almost equal balance between sch vs ctrl
+            ids = shuffle(ids, random_state=0, n_samples=n_subjects)
```
Comment (AlexandreAbraham) [inline nilearn/datasets/func.py]: Isn't that a bit dangerous for reproducibility?

File: nilearn/datasets/func.py [hunk 5]
```diff
@@ -1411,3 +1412,98 @@ def fetch_megatrawls_netmats(dimensionality=100, timeseries='eigen_regression',
...
+    func_filenames = [('fmri_' + i.decode().strip(' "\'') +
+                       '_session1' + '_run1.nii.gz') for i in ids]
+    mats_filenames = [('fmri_' + i.decode().strip(' "\'') +
+                       '_session1' + '_run1_extra.mat') for i in ids]
+
+    func_files = [(path, url, opts) for path in func_filenames]
+    mat_files = [(path, url, opts) for path in mats_filenames]
+
+    func_files = _fetch_files(data_dir, func_files, verbose=verbose)
+    mat_files = _fetch_files(data_dir, mat_files, verbose=verbose)
```
Comment (AlexandreAbraham) [inline nilearn/datasets/func.py]: `mat` files cannot be processed by nilearn. Should we load them or add some code to process them automatically (can be another PR).

File: nilearn/datasets/description/cobre.rst (hunk 2 above)
Comment (KamalakerDadi) [inline nilearn/datasets/description/cobre.rst]: Yes, but as Pierre said in the issue that it will take time and we decided to proceed with this version.

File: nilearn/datasets/func.py (hunk 3 above)
Comment (KamalakerDadi) [inline nilearn/datasets/func.py]: I didn't get your comment. I thought you understand the situation to download each individual file.

File: nilearn/datasets/func.py (hunk 4 above)
Comment (KamalakerDadi) [inline nilearn/datasets/func.py]: which one ```random_state=0``` ?

File: nilearn/datasets/description/cobre.rst (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/description/cobre.rst]: OK!

File: nilearn/datasets/func.py (hunk 3 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/func.py]: Well, Gaël said in the issue that a 1 hour download was too much. Plus it does not make sense to return 10 subjects by default if you force downloading the whole dataset. But nevermind, let's move on.

File: nilearn/datasets/func.py (hunk 4 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/func.py]: `random_state=0` may give different results depending on the machine used. If subjects are splitted controls/schizophrenic, maybe we should just take the first `n_subjects/2` controls and the first `n_subject/2` schizos?

File: nilearn/datasets/func.py (hunk 5 above)
Comment (KamalakerDadi) [inline nilearn/datasets/func.py]: > Should we load them or add some code to process them automatically (can be another PR).

I will keep it open to matlab users for example. I don't think we need to load them.

File: nilearn/datasets/func.py (hunk 4 above)
Comment (KamalakerDadi) [inline nilearn/datasets/func.py]: Done. Can you see it is Ok ?

**Comment by GaelVaroquaux** [comment]
LGTM. :+1: for merge
